packages/client/src/internals/dataLoader.ts

```diff
@@ -26,100 +37,125 @@ type BatchLoader<TKey, TValue> = {
 export function dataLoader<TKey, TValue>(
   batchLoader: BatchLoader<TKey, TValue>,
 ) {
-  let batch: Batch<TKey, TValue> | null = null;
-  let dispatchTimer: NodeJS.Timer | number | null = null;
+  let pendingItems: BatchItem<TKey, TValue>[] | null = null;
+  let dispatchTimer: ReturnType<typeof setTimeout> | null = null;
 
-  const destroyTimerAndBatch = () => {
+  const destroyTimerAndPendingItems = () => {
     clearTimeout(dispatchTimer as any);
     dispatchTimer = null;
-    batch = null;
+    pendingItems = null;
   };
-  function dispatch() {
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    const thisBatch = batch!;
-    destroyTimerAndBatch();
 
-    const itemBatches: BatchItem<TKey, TValue>[][] = [[]];
+  /**
+   * Iterate through the items and split them into groups based on the `batchLoader`'s validate function
+   */
+  function groupItems(items: BatchItem<TKey, TValue>[]) {
+    const groupedItems: BatchItem<TKey, TValue>[][] = [[]];
+    let index = 0;
+    while (true) {
+      const item = items[index];
+      if (!item) {
+        // we're done
+        break;
+      }
+      const lastGroup = groupedItems[groupedItems.length - 1];
+
+      if (item.aborted) {
+        // Item was aborted before it was dispatched
+        item.reject(new Error('Aborted'));
+        index++;
+        continue;
+      }
 
-    for (const item of thisBatch.items) {
-      const lastItemBatch = itemBatches[itemBatches.length - 1];
       const isValid = batchLoader.validate(
-        lastItemBatch.concat(item).map((_item) => _item.key),
+        lastGroup.concat(item).map((it) => it.key),
       );
+
       if (isValid) {
-        lastItemBatch.push(item);
-      } else {
-        itemBatches.push([item]);
-        if (lastItemBatch.length === 0) {
-          itemBatches.push([]);
-        }
+        lastGroup.push(item);
+        index++;
+        continue;
+      }
+
+      if (lastGroup.length === 0) {
+        item.reject(new Error('Input is too big for a single dispatch'));
+        index++;
+        continue;
       }
+      // Create new group, next iteration will try to add the item to that
+      groupedItems.push([]);
     }
+    return groupedItems;
+  }
+
+  function dispatch() {
+    const groupedItems = groupItems(pendingItems!);
+    destroyTimerAndPendingItems();
 
-    const promises: Promise<void>[] = [];
-    const cancels: CancelFn[] = [];
+    // Create batches for each group of items
+    for (const items of groupedItems) {
+      if (!items.length) {
+        continue;
+      }
+      const batch: Batch<TKey, TValue> = {
+        items,
+        cancel: throwFatalError,
+      };
+      for (const item of items) {
+        item.batch = batch;
```

Comment:
We don't attach the item to the batch until here, if it's cancelled before here it'll be omitted by the grouping